### ISSUE TITLE
`drjit.grad` always return 0 after first updating of params

### ISSUE THREAD

**Issue by grkid** [mitsuba-renderer/mitsuba3#218]
I was trying to integrate mitsuba3 with pyTorch manually, so i write some code in order to find out how to fetch gradients according to Gradient-based optimization tutorial.
```python
import drjit as dr
import mitsuba as mi
import numpy as np
import torch

def mse(image):
    return dr.mean(dr.sqr(image - image_ref))

if __name__=="__main__":
    mi.set_variant('cuda_ad_rgb')
    scene = mi.load_file('cbox.xml', res=128, integrator='prb')

    image_ref = mi.render(scene, seed=0, spp=512)
    mi.util.convert_to_bitmap(image_ref)
    mi.util.write_bitmap("ref.exr",image_ref)

    # traverse
    params = mi.traverse(scene)

    key = 'red.reflectance.value'

    # Save the original value
    param_ref = mi.Color3f(params[key])

    # Set another color value and update the scene
    params[key] = mi.Color3f(0.01, 0.2, 0.9)
    params.update()

    image_init = mi.render(scene, seed=0, spp=128)
    mi.util.convert_to_bitmap(image_init)

    params[key] = mi.Color3f(0.01, 0.2, 0.9)
    params.update()


    learnRate=2.0
    errors = []
    for it in range(1000):

        dr.enable_grad(params[key])
        params.update()
        image = mi.render(scene, params, spp=4)
        loss = mse(image)
        dr.backward(loss)
        paramGrad=dr.grad(params[key])
        # update params according to paramGrad
        params[key]=mi.Color3f(params[key])-mi.Color3f(paramGrad)*learnRate
        params.update()

        # Track the difference between the current color and the true value
        err_ref = dr.sum(dr.sqr(param_ref - params[key]))
        print(f"Iteration {it:02d}: loss = {loss[0]}, param err = {err_ref[0]} param = {params[key]}", end='\r')
        errors.append(err_ref)
    print('\nOptimization complete.')

    image_final = mi.render(scene, seed=1,spp=512)
    mi.util.convert_to_bitmap(image_final)
    mi.util.write_bitmap("final.exr",image_final)
```

BUT after first updating of `params[key]` (it >=1), `paramGrad` is always like this:
![image](https://user-images.githubusercontent.com/46410388/187187965-5fc4bc47-03e7-4e98-acd4-ddb5a2bb6eb4.png)

I managed to solve this by replacing
`params[key]=mi.Color3f(params[key])-mi.Color3f(paramGrad)*learnRate`
to
`params[key]=mi.Color3f(params[key][0][0]-learnRate*paramGrad[0][0], params[key][1][0]-learnRate*paramGrad[1][0], params[key][2][0]-learnRate*paramGrad[2][0])`
and it works (like magic). I don't know why so I guess it is a bug to be fixed. And, write such code could introduce more bugs, so I want to know if there exists any other way to update params easily. Any assistance will be appreciated.

**Comment by Speierers** [comment]
You will need to call `dr.enable_grad(params[key])` after updating your parameters to make them differentiable again (this is how it is implemented in `optimizers.py`).

I am not enterily sure why your other implementation works, but it definitely doesn't look right ;)

**Comment by grkid** [comment]
```python
for it in range(1000):

        dr.enable_grad(params[key])
        params.update()
        image = mi.render(scene, params, spp=4)
        loss = mse(image)
        dr.backward(loss)
        paramGrad=dr.grad(params[key])
        params[key]=mi.Color3f(params[key])-mi.Color3f(paramGrad)*learnRate
        params.update()
        dr.enable_grad(params[key])

        # Track the difference between the current color and the true value
        err_ref = dr.sum(dr.sqr(param_ref - params[key]))
        print(f"Iteration {it:02d}: loss = {loss[0]}, param err = {err_ref[0]} param = {params[key]}", end='\r')
        errors.append(err_ref)
```
and it still doesn't work.
![image](https://user-images.githubusercontent.com/46410388/187192350-5adefeef-ff0f-4dc3-b3fc-db3e83578705.png)

**Comment by Speierers** [comment]
Sorry, I meant that you need to call this after the dictionary update, but before the `params.update()`.
```
params[key]=mi.Color3f(params[key])-mi.Color3f(paramGrad)*learnRate
dr.enable_grad(params[key])
params.update()
```

**Comment by grkid** [comment]
as always ... 🥺
![image](https://user-images.githubusercontent.com/46410388/187193753-cb9c6d94-f626-4767-8072-a05d81503cbd.png)

**Comment by Speierers** [comment]
This is strange. I would try to get this script to work using one of the `Optimizer` class, e.g. `SGD`. Once this works you can break the optimizer appart and implement your own (like you are trying to do here).

**Comment by grkid** [comment]
Thank you so much. It will be very helpful to me if you could inform me later as i need to use it recently.

**Comment by orbitchen** [comment]
Well, as the torch integration of mitsuba3 has been implemented now, I suppose that the gradients should not be updated manually. You may refer to the official documentation.

**Comment by WhyICanNotDo** [comment]
Thank you for your prompt response. I followed the official documentation to write the code, but the gradient is still 0. There might be an issue with my code in another module.
